Feature: CSR Table Block

stephiescastle opened the pull request:
### Checklist

- [x] Include a description of your pull request and instructions for the reviewer to verify your work.
- [x] Link to the issue if this PR is issue-specific.
- [x] Create/update the corresponding story if this includes a UI component.
- [ ] Create/update documentation. If not included, tell us why.
- [x] List the environments / browsers in which you tested your changes.
- [x] Tests, linting, or other required checks are passing.
- [x] PR has an informative and human-readable title
  - PR titles are used to generate the change log in [releases](../releases); good ones make that easier to scan.
  - PRs will be broadly categorized in the change log, but for even easier scanning, consider prefixing with a component name or other useful categorization, e.g., "BaseButton: fix layout bug", or "Storybook: Update dependencies".
- [x] PR has been tagged with a [SemVer](https://semver.org/) label and a [general category label](https://github.com/nasa-jpl/explorer-1/blob/52994b671411f55961d7beb8851d4580ac3f434f/.github/release-drafter.config.yml#L21-L39), or skip-changelog.
  - These tags are used to do the aforementioned broad categorization in the change log and determine what the next release's version number should be.
  - [Release Drafter](https://github.com/marketplace/actions/release-drafter) will attempt to do the category labeling for you! Please double-check its work.

## Description

Adds a new block "CsrTableBlock" to support displaying large amounts of data provided by a CSR endpoint. https://github.com/nasa-jpl/www/issues/817

This block can be later adapted to support a more generalized interactive/filterable table for Explorer-1.

![image](https://github.com/user-attachments/assets/24eed49d-0bd7-476d-bb83-d68a139dd37c)

## Instructions to test

1. `pnpm i`
2. `make vue-storybook`
3. View the story: http://localhost:6006/?path=/story/components-featureflags-blockcsrtable--base-story
4. Test a custom API endpoint by populating the `apiEndpoint` prop under the "Controls" tab. Contact me for the endpoint.

## Tested in the following environments/browsers:

<!-- Delete this section if not applicable. -->

### Operating System

- [x] macOS
- [ ] iOS
- [ ] iPadOS
- [ ] Windows

### Browser

- [x] Chrome
- [x] Firefox ESR
- [ ] Firefox
- [x] Safari
- [ ] Edge

Scotchester commented on packages/vue/src/components/BlockCsrTable/BlockCsrTable.stories.js:
```diff
@@ -0,0 +1,220 @@
+      element: 'body',
+      config: {
+        rules: [
+          // ignore rules that fail tests but are not actually violations
+          // see: https://github.com/ag-grid/ag-grid/issues/6218
+          {
+            id: 'aria-required-children',
+            enabled: false
+          },
+          {
+            id: 'region',
+            enabled: false
+          }
```
It also doesn't bubble up when included in other stories. The errors appear in the UI on `BlockStreamfield` and in any story that includes it.

stephiescastle replied:
yeah, i'm still trying to figure out what's going on here. Pretty frustrating that the test-runner and accessibility tab behave differently.

Scotchester commented on packages/vue/src/components/BlockCsrTable/BlockCsrTable.stories.js:
```diff
@@ -0,0 +1,220 @@
+            enabled: false
+          },
+          {
+            id: 'region',
```
I had to change this to `'scrollable-region-focusable'` to get that error to go away, but then the actual `'region'` error ("All page content should be contained by landmarks") started appearing, and I'm not sure why that error doesn't appear on every other story, since it's never been disabled before.

Scotchester reviewed (commented):
This is pretty neat! I do have a few minor concerns, though.

I'm not sure about the dynamic height. It makes the filter options pop-up a bit squirrelly when the overall height is very short

![filter pop up with scrollbars](https://github.com/user-attachments/assets/1cc57cb3-a3f9-4467-b14d-1884a1f4015a)

and the layout shift seems less than ideal, especially the UX of the bottom-of-table controls moving all about.

Another issue I came across: In the table within the modal, the filter pop-up doesn't have the Reset and Apply buttons. I know that Apply is sort of a placebo, since they're applied automatically, but not being able to reset them is an issue.

![Table in modal with filter pop-up activated and scrolled all the way down](https://github.com/user-attachments/assets/0870422b-59eb-44de-99dc-2508f8fb2913)

stephiescastle commented:
> I'm not sure about the dynamic height. It makes the filter options pop-up a bit squirrelly when the overall height is very short 

Yeah, I couldn't find the right balance for this. The table could potentially have hundreds of rows, with the option to display 20, 50 or 100 on them. It will also be embedded on a page. So if we have a set height, there is bound to be some scrolling within the table, or a ton of white space at the bottom of the page. :/

> In the table within the modal, the filter pop-up doesn't have the Reset and Apply buttons.

I had to configure these to appear, and I forgot to do it for the nested table! I'll add that.

Scotchester commented:
> Yeah, I couldn't find the right balance for this. The table could potentially have hundreds of rows, with the option to display 20, 50 or 100 on them. It will also be embedded on a page. So if we have a set height, there is bound to be some scrolling within the table, or a ton of white space at the bottom of the page. :/

Couldn't we avoid scrolling _within_ the table if the height is just always based on the number-of-rows option?

stephiescastle commented:
> > I'm not sure about the dynamic height. It makes the filter options pop-up a bit squirrelly when the overall height is very short
> 
> Yeah, I couldn't find the right balance for this. The table could potentially have hundreds of rows, with the option to display 20, 50 or 100 on them.

I'm going to see if I can change the height based on the results per page selection, and then remove dynamic height.

stephiescastle commented:
@Scotchester I added behavior that sets a min-height based on the page size. It seems a bit much, especially when trying 50 and 100. I'm thinking maybe we just set a min height that matches the default setting (20 items), and not worry about the larger pages?

Also, note that I can't get the heights to be 100% exact, as there will be text wrapping in the table.

I also fixed the behavior of the filters in the nested table (it now has apply and reset buttons)

Scotchester commented:
> @Scotchester I added behavior that sets a min-height based on the page size. It seems a bit much, especially when trying 50 and 100. I'm thinking maybe we just set a min height that matches the default setting (20 items), and not worry about the larger page sizes?

Yeah, seeing it now, I think I agree. Whatever height you think makes sense.

> I also fixed the behavior of the filters in the nested table (it now has apply and reset buttons)

I see that, but I'm now having an issue with filtering in the nested table. (It may have also happened before the latest commits; I'm not sure.) This filter is applied and not correctly including the item in the resulting set:

![Filter set to equals 200 isn't returning item with 200](https://github.com/user-attachments/assets/b2399e50-2e2d-4b92-971b-c40d0433f3df)

The less than filter is also off by 1:

![Filter set to less than 200 still returns item with 200](https://github.com/user-attachments/assets/0c9fb0b9-83d0-4ad1-af9f-b4fd61c05527)

stephiescastle commented:
@Scotchester I've fixed the filter types in the nested table. I had to switch them to `string` filters, as sometimes the results won't always be numeric. I've reviewed this with the CSR Team, and they're in agreement.

I still haven't gotten further with the accessibility tests. If I can't work it out today, it might be something we have to figure out after the fact :/

stephiescastle commented:
I cleaned up the tests as much as I could. I'm not sure how to get the a11y ignore rules to work in the test-runner. It's also been difficult to debug as I get very different test results locally.

In ci:

```
[TEST] Test Suites: 2 failed, 2 skipped, 158 passed, 160 of 162 total
[TEST] Tests:       3 failed, 3 skipped, 455 passed, 461 total
```

Compared to locally:

```
Test Suites: 8 failed, 2 skipped, 152 passed, 160 of 162 total
Tests:       21 failed, 3 skipped, 437 passed, 461 total
```

Scotchester reviewed (approved):
Works for me!
